kaggle_environments/envs/lux_ai_2022/luxai2022/env.py

```diff
@@ -0,0 +1,618 @@
+        self.max_episode_length = self.env_cfg.max_episode_length
+        for agent in self.possible_agents:
+            self.state.units[agent] = OrderedDict()
+            self.state.factories[agent] = OrderedDict()
+        obs = self.state.get_obs()
+        observations = {agent: obs for agent in self.agents}
+        return observations
+
+    def step(self, actions):
+        """
+        step(action) takes in an action for each agent and should return the
+        - observations
+        - rewards
+        - dones
+        - infos
+        dicts where each dict looks like {agent_1: item_1, agent_2: item_2}
+        """
+        # If a user passes in actions with no agents, then just return empty observations, etc.
+        if not actions:
+            raise ValueError("No actions given")
+            self.agents = []
+            return {}, {}, {}, {}
+
+        failed_agents = {agent: False for agent in self.agents}
+        # Turn 1 logic, handle # TODO Bidding
+        early_game = False
+        if self.env_steps == 0:
+            early_game = True
+        if self.env_cfg.BIDDING_SYSTEM and self.env_steps <= self.state.board.factories_per_team + 1:
+            early_game = True
+
+        if early_game:
+            # handle initialization
+
+            if self.env_cfg.BIDDING_SYSTEM:
+                if self.env_steps == 0:
+                    highest_bid = -1
+                    highest_bid_agent = None
+                    for k, a in actions.items():
+                        if a is None:
+                            failed_agents[k] = True
+                            continue
+                        if k not in self.agents:
+                            raise ValueError(f"Invalid player {k}")
+                        if "faction" in a and "bid" in a:
+                            self.state.teams[k] = Team(
+                                team_id=self.agent_name_mapping[k], agent=k, faction=FactionTypes[a["faction"]]
+                            )
+                            self.state.teams[k].init_water = self.env_cfg.INIT_WATER_METAL_PER_FACTORY * (self.state.board.factories_per_team)
+                            self.state.teams[k].init_metal = self.env_cfg.INIT_WATER_METAL_PER_FACTORY * (self.state.board.factories_per_team)
+                            self.state.teams[k].factories_to_place = self.state.board.factories_per_team
+                            # verify bid is valid
+                            valid_action = True
+                            bid = a["bid"]
+                            if bid < 0 or bid > self.state.teams[k].init_water:
+                                valid_action = False
+                            if not valid_action:
+                                failed_agents[k] = True
+                                continue
+                            if bid > highest_bid:
+                                highest_bid = bid
```

Comment:
we talked about having to pay the cost of the 2nd highest bid, but I think that could lead to meta gaming of:

top agents both bid "max possible water" (where that value is generally agreed upon and they all tie
against anyone else they win the factory for cheaper than they would have paid

Reply:
yeah this is exactly the scenario I considered. Left it as bidding what you bid for